### PR TITLE
Feature/KAK-9-checkers

### PR DESCRIPTION
**Pull request by simonsipl**

### CODE REVIEW

File: src/step_definitions/elements.ts [hunk 1]
```diff
@@ -188,180 +93,35 @@ Then(/^there are "([^"]*)" dropdown list elements with following options:$/, fun
...
 });
 
-Then(/^there is no element "([^"]*)" with value "([^"]*)"$/, function(elementName, value) {
-  const pageElement = this.currentPage.getElement(elementName);
-
-  return matchers
-    .match(pageElement, valueToTextTransformer.transform(value))
-    .catch(() => Promise.resolve(false))
-    .then(result => (result ? Promise.reject() : Promise.resolve()));
+Then(/^there (is|is no) element "([^"]*)" containing "([^"]*)" text$/, function(assert, elementName, value) {
```
Comment (tgorskievo) [inline src/step_definitions/elements.ts]: Did you checked if autosuggestion works for this step?

File: src/web/methods/checkers.methods.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+              );
+            }
+          }
+        })
+        .then(() => Promise.all(promises));
+    });
+  });
+};
+
+export const checkFollowingElementsForElement = (
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: Improve the names of those methods - I don't get what this method is used for

File: src/step_definitions/elements.ts (hunk 1 above)
Comment (simonsipl) [reply src/step_definitions/elements.ts]: yes. It gives you a suggestion template `And there is|is no element "<string>" containing "<string>" text`

File: src/web/methods/checkers.methods.ts (hunk 2 above)
Comment (simonsipl) [reply src/web/methods/checkers.methods.ts]: @tgorskievo @joszo suggested to remove check from this names as we use `methods.checkers.nameOfCheck`

Comment (sethii) [reply src/web/methods/checkers.methods.ts]: I dont see a reason to remove "check". I suppose this is a method to check if array of items is available under specific selector right?

Comment (simonsipl) [reply src/web/methods/checkers.methods.ts]: yes @sethii, I was using step definitions to name those export functions so they, in my opinion, are straight forward

File: src/web/methods/checkers.methods.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+import { matchers } from '../../matchers';
+import { createValueToTextTransformer } from '../../transformers/transformer/values-to-text.transformer';
+import { TableDefinition } from 'cucumber';
+import * as chai from 'chai';
+import { methods } from './index';
+import variableStore from '../../core/variable-store.helper';
+
+const valueToTextTransformer = createValueToTextTransformer();
+
+export const checkNumberOfElements = (currentPage: BasePage, numberExpression: string, element: string) => {
```
[the remark this answers is not in the document]
Comment (simonsipl) [reply src/web/methods/checkers.methods.ts]: Fixed

File: src/web/methods/checkers.methods.ts [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+  const numbers = numberMatches.map(item => parseInt(item));
+
+  const expectFunction = async (words, num) => {
+    const numberOfElements = await currentPage.getNumberOfElements(elementName);
+    return chai.expect(numberOfElements).to.be[words.pop()](...num);
+  };
+  return expectFunction(numberExpression.substr(0, numberExpression.indexOf(numberMatches[0]) - 1).split(' '), numbers);
+};
+
+export const checkElementsForElement = (currentPage: BasePage, elementName: string, data: TableDefinition) => {
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementsIncludeElement

File: src/web/methods/checkers.methods.ts [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+  return currentPage.waitForVisibilityOf(elementName).then(() => {
+    const promises = checkers.map(check => {
+      return matchers.match(currentPage.getElement(check.element), valueToTextTransformer.transform(check.matcher));
+    });
+
+    return Promise.all(promises);
+  });
+};
+
+export const checkElementWithValue = (currentPage: BasePage, assert: string, elementName: string, value: string) => {
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementContainsValue

File: src/web/methods/checkers.methods.ts [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+  } else if (assert === 'is not') {
+    return currentPage.waitForVisibilityOf(elementName).then(() => {
+      return checkMatcher(pageElement, value)
+        .catch(() => Promise.resolve(false))
+        .then(result => (result ? Promise.reject() : Promise.resolve()));
+    });
+  }
+};
+
+export const checkElementWithRegex = (currentPage: BasePage, assert: string, elementName: string, matcher: string) => {
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementMatchesRegex

File: src/web/methods/checkers.methods.ts [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+  } else if (assert === 'is not') {
+    return currentPage.waitForVisibilityOf(elementName).then(() => {
+      return checkMatcher(pageElement, matcher)
+        .catch(() => Promise.resolve(false))
+        .then(result => (result ? Promise.reject() : Promise.resolve()));
+    });
+  }
+};
+
+export const checkElementWithMatcher = (
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementMatchesMatcher

File: src/web/methods/checkers.methods.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+  } else if (assert === 'is not') {
+    return currentPage.waitForVisibilityOf(elementName).then(() => {
+      return checkMatcher(pageElement, matcher)
+        .catch(() => Promise.resolve(false))
+        .then(result => (result ? Promise.reject() : Promise.resolve()));
+    });
+  }
+};
+
+export const checkElementContainingText = (
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementContainsText

File: src/web/methods/checkers.methods.ts [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+  } else if (assert === 'is not') {
+    return currentPage.waitForVisibilityOf(elementName).then(() => {
+      return checkMatcher(pageElement, value)
+        .catch(() => Promise.resolve(false))
+        .then(result => (result ? Promise.reject() : Promise.resolve()));
+    });
+  }
+};
+
+export const checkFollowingElementsInTable = (currentPage: BasePage, table: string, data: TableDefinition) => {
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkElementsForValues

File: src/web/methods/checkers.methods.ts [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+              );
+            }
+          }
+        })
+        .then(() => Promise.all(promises));
+    });
+  });
+};
+
+export const checkFollowingElementsForElement = (
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementsIncludeElement

File: src/web/methods/checkers.methods.ts [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+              )
+            );
+          });
+        })
+        .then(() => Promise.all(promises));
+    });
+  });
+};
+
+export const checkElementsHaveSameValueForElement = (
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfAllElementsHaveMatchingValues

File: src/web/methods/checkers.methods.ts [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+            .getText()
+            .then(elementText => {
+              expect(elementText).toEqual(firstElementText);
+            });
+        });
+      });
+  });
+};
+
+export const checkElementHaveItemsWithValue = (
```
Comment (tgorskievo) [inline src/web/methods/checkers.methods.ts]: checkIfElementHaveItemsWithValue

File: src/step_definitions/elements.ts [hunk 13]
```diff
@@ -79,91 +60,20 @@ When(/^I store table "([^"]*)" rows as "([^"]*)" with columns:$/, function(table
...
-                element.element(self.currentPage.getElement(hash[0]).locator()),
-                valueToTextTransformer.transform(hash[1])
-              )
-            );
-          });
-        })
-        .then(() => Promise.all(promises));
-    });
-  });
+  return methods.checkers.checkTableElementContainMatchingNumberOfElement(
```
Comment (tgorskievo) [inline src/step_definitions/elements.ts]: `checkIfTableElementContainsMatchingNumberOfElements`

File: src/step_definitions/elements.ts [hunk 14]
```diff
@@ -384,7 +149,7 @@ Then(/^"([^"]*)" value on the "([^"]*)" list is sorted in "([^"]*)" order$/, fun
 });
 
 When(/^I infinitely scroll to the "([^"]*)" element$/, function(elementName) {
-  return methods.interactions.scrollToLoader(this.currentPage, elementName);
+  return methods.interactions.infinityScroll(this.currentPage, elementName);
```
Comment (tgorskievo) [inline src/step_definitions/elements.ts]: `infinityScrollTo`

File: src/step_definitions/elements.ts (hunk 13 above)
Comment (simonsipl) [reply src/step_definitions/elements.ts]: Done

File: src/step_definitions/elements.ts (hunk 14 above)
Comment (simonsipl) [reply src/step_definitions/elements.ts]: Done

**Comment by tgorskievo** [comment]
Please squash those commits :)

File: package-lock.json [hunk 15]
```diff
@@ -3565,9 +3565,15 @@
       "integrity": "sha1-8QdIy+dq+WS3yWyTxrzCivEgwIE="
     },
     "handlebars": {
+<<<<<<< HEAD
```
Comment (tgorskievo) [inline package-lock.json]: Not resolved conflict 😈

**Pull request merged**